### PR TITLE
[Merged by Bors] - feat(RingTheory): valuation subrings are maximal local subrings

### PR DESCRIPTION
**Pull request by erdOne**
from the valuative criterion project

Co-authored-by: Yaël Dillies
Co-authored-by: Javier López-Contreras

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
I feel like it should be fine to import `MaximalIdeal.Basic` into `LocalRing.RingHom.Basic` but I can also inline the proof used to avoid the import.

File: Mathlib/RingTheory/LocalRing/RingHom/Basic.lean [hunk 1]
```diff
@@ -116,6 +116,14 @@ theorem of_surjective [CommSemiring R] [IsLocalRing R] [CommSemiring S] [Nontriv
       (isUnit_or_isUnit_of_isUnit_add <| IsLocalHom.map_nonunit _ hab).imp f.isUnit_map
         f.isUnit_map)
 
+lemma _root_.IsLocalHom.of_surjective [CommRing R] [CommRing S] [Nontrivial S] [IsLocalRing R]
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalRing/RingHom/Basic.lean:119]: Or maybe `Function.Surjective.isLocalHom`?
```suggestion
lemma _root_.Function.Surjective.isLocalHom [CommRing R] [CommRing S] [Nontrivial S] [IsLocalRing R]
```
What do you think?

File: Mathlib/RingTheory/LocalRing/Subring.lean [hunk 2]
```diff
@@ -0,0 +1,286 @@
+/-
+Copyright (c) 2024 Andrew Yang, Yaël Dillies, Javier López-Contreras. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang, Yaël Dillies, Javier López-Contreras
+-/
+import Mathlib.RingTheory.Ideal.Over
+import Mathlib.RingTheory.Valuation.ValuationSubring
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalRing/Subring.lean:7]: This import feels like it shouldn't appear in a file with this name.
Can you please split the current file into `LocalRing/Subring` and `LocalRing/Valuation`?

File: Mathlib/RingTheory/LocalRing/RingHom/Basic.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/RingTheory/LocalRing/RingHom/Basic.lean:119]: I added an alias.

File: Mathlib/RingTheory/LocalRing/Subring.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/RingTheory/LocalRing/Subring.lean]: I moved the `isMax` results to `Valuation/LocalSubring` (ideally it should live in `Valuation/ValuationSubring` but that file is already quite long)

File: Mathlib/RingTheory/LocalRing/Subring.lean [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+def range [IsLocalRing R] [Nontrivial S] (f : R →+* S) : LocalSubring S :=
+  .copy (map f (mk ⊤)) f.range (by ext x; exact congr(x ∈ $(Set.image_univ.symm)))
+
+/--
+The domination order on local subrings.
+`A` dominates `B` if and only if `B ≤ A` and `m_A ∩ B = m_B`.
+-/
+@[stacks 00I9]
+instance : PartialOrder (LocalSubring R) where
+  le A B := ∃ h : A.1 ≤ B.1, IsLocalHom (Subring.inclusion h)
```
Comment (jcommelin) [inline Mathlib/RingTheory/LocalRing/Subring.lean:70]: Could you please add a lemma `LocalSubring.le_iff` that allows to rewrite `A \le B` into this definition, and please put a similar docstring on that lemma as on this instance.

**Comment by jcommelin** [comment]
bors d+

**Comment by erdOne** [comment]
Thanks for the review!
bors merge

**Pull request closed without merging**